xwiki-platform-core/xwiki-platform-security/xwiki-platform-security-authorization/xwiki-platform-security-authorization-api/src/main/java/org/xwiki/security/authorization/cache/internal/DefaultSecurityCache.java

```diff
@@ -131,12 +141,14 @@ private Cache<SecurityCacheEntry> newCache() throws InitializationException
     public void initialize() throws InitializationException
     {
         cache = newCache();
+        this.internalEntries = new ReferenceMap<>(AbstractReferenceMap.ReferenceStrength.HARD,
+            AbstractReferenceMap.ReferenceStrength.WEAK);
```

Comment:
Isn't this an open door for memory leak: I mean if the cache reach a given size, entries are automatically removed, but the map will still keep the entries no? So isn't there a risk for this map to grow indefinitely? 

You said in the javadoc:
> When an entry is removed from the cache, the GC will also remove it from the internal entries unless it is still referenced as a parent somewhere else.

So it means it could be the case no? Unless you assume that the risk is very low that a lot of references are kept without being in the cache anymore?

Reply:
I recommend reading the documentation of the [security inspector application](https://extensions.xwiki.org/xwiki/bin/view/Extension/Security+Inspector+Application) as I think it is the best description that we have of the contents of the security cache. Ignoring some special cases, we have the following cases:

1. Security access entries. They store the outcome of a right computation for a specific user/document combination. Those are never the parent of another entry so they should always be collected by the GC when they are disposed from the actual cache.
2. Entries that are representing regular documents, spaces or wikis. They store the access rules that are associated with the document/space/wiki. They can be the parent of another document, space, access entry, user or group. In any case, they must have some descendant that is actually in the size-limited cache and the number of documents above them is limited by the depth of the hierarchy. Further, their number is naturally limited by the number of documents in the XWiki installation. If you have deep hierarchies and the cache is just storing an access entry, you could indeed have, say, 5 times as many entries in the internal map than in the size-constrained cache. That said, any computation of a security access entry accesses the whole hierarchy and thus tries their re-insertion in the size-constrained cache. Further, in the UI, we frequently display the whole path to a document and thus check the whole hierarchy. So I think the chances that you have a long hierarchy without many entries in the size-constrained cache are low.
3. Entries that are representing users or groups. These are basically the same as documents but additionally they have parent groups as parents. Again, I don't expect huge hierarchies there. Also, we don't store all members of a group so even huge groups are not an issue.

Reply:
To provide some context why all of this work is necessary and why having some more entries in memory might be good:

Imagine that we have a wiki that has primarily public access. Ideally, almost all security access entries for public access and all security access entries for author right checks should be in the security cache. Therefore, only these are accessed and nothing of the hierarchy behind them. Imagine now that the size of the security cache is not enough to store all of them, so we just store the most popular ones. Now the cache needs to decide which ones to keep. The cache actually has access frequency data, at least as an approximate sketch. So it can make informed decisions that these entries that are representing regular documents are never accessed and it will remove them. Currently, this will also remove the frequently accessed security access items that have them as parents. With this change, instead, these "internal" entries will remain in the internal map and the frequently accessed security access items can remain in the cache as the hierarchy is intact (if the document is modified, it needs to be removed from the cache including all its children, for that purpose we need to be able to find it).

So in this scenario we will indeed store more than originally configured, but I think this is actually good.